### PR TITLE
Jetty 12 - Issue #8474 - return null on resolve() if resource doesn't exist

### PR DESCRIPTION
**Pull request by joakime**
* makes `Resource.resolve(String)` return null if the resulting `Resource` would not exist.
* makes `ResourceFactory.newResource()` (and similar methods) return null if the resulting `Resource` would not exist.

### CODE REVIEW

**Comment by joakime** [comment]
@sbordet here's the minimal change to have PathResource.resolve() behave the same as ResourceCollection.resolve().

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:280]: I also think we have rewrite rules for removing // in user supplied URIs.   I really don't think we need to do this in resolve and I'd like to see what tests break if we remove it.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:284]: This is probably needed.... but I'd still be interested to know exactly why.  i.e what fails if we don't do this.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:259]: I like having a resolve in PathResource rather than try to be generic in Resource

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: Can it be abstract?

**Review by gregw — CHANGES_REQUESTED** [review]
This is a bit of an omnibus PR.  There are too many changes in it that are unrelated to the premise (of a null return).

There is goodness in this PR, but it is too hard to review when there are 70 files changed with renames, reformatting, various fixes and cleanups.

Can you break this out into multiple PRs, so that ultimately there is a single PR with the null return change.

**Comment by joakime** [comment]
Thank you for the preliminary review.

For now, hold off on any further reviews until I can move some of the oddball changes out of this PR into separate PRs (as once those are merged, it will cause a rebase in this PR to clean out those changes)

**Comment by joakime** [comment]
The removal of `exist()` has not been done on this PR.

**Comment by gregw** [comment]
> Thank you for the preliminary review.
> 
> For now, hold off on any further reviews until I can move some of the oddball changes out of this PR into separate PRs (as once those are merged, it will cause a rebase in this PR to clean out those changes)

For anything like renaming of variables / fields etc.   I think you can probably do them without a PR if they are not part of the public API.  Just check with another developer and if they are OK, push directly to the branch.

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceTest.java
Comment (joakime) [inline jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceTest.java:262]: Need to revisit this

**Comment by joakime** [comment]
From latest run https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-8597/10/tests

**org.eclipse.jetty.tests.distribution.DemoModulesTests.testJspDump([1] ee8) Test Failure**

```
java.lang.IllegalArgumentException: KeyStore Path does not exist: /home/jenkins/agent/workspace/jetty.project_PR-8597
  /tests/test-distribution/test-distribution-common/target/tests/test-bases/jetty_base_5828043306936187437/etc/test-keystore.p12
```

I wonder if this should be throwing, or not.
I'm leaning towards throwing.  (and fixing the testcase to actually have a keystore)

@sbordet and @lorban what do you think?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java:129]: I think more changes below  are needed for code clarity.

specifically if resources == null, can we just explicitly return null?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java
Comment (lorban) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java:229]: We are not adding `null`s into `_monitored` for non-existing resources, so this null check can go away.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java
Comment (lorban) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java:345]: Why adding the `resources` collection for each resource?

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java
Comment (lorban) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:730]: The `Path` may require mounting, right? So this may lead to a mount leak if I'm not mistaken.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/MountedPathResource.java
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/MountedPathResource.java:31]: Is `public` needed?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:50]: Is `public` needed?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/MountedPathResource.java
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/MountedPathResource.java:47]: Is the other constructor still needed?

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:280]: This is addressed in PR #8631

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:284]: There is a behavior different with regards what is remembered by `Path` when you use `Paths.get(URI)` or `Paths.of(URI)` vs `Paths.of(String)`.
Path segments in a URI like `/../` are remembered and will impact behavior like `Files.exists(path)`.
This seems to be because a URI is seen as canonical and the raw `/../` is passed all the way down to the real filesystem.
Be happy to demonstrate with you later.

**Comment by joakime** [comment]
Closing this large (in commit count) PR in favor of #8702

**Pull request closed without merging**
